base_layer/wallet/src/output_manager_service/service.rs

```diff
@@ -748,6 +759,110 @@ where
+        let mut estimated_fee = MicroTari::default();
+        let mut utxo_total = MicroTari::default();
+        loop {
+            estimated_fee = Fee::calculate(fee_per_gram, 1, estimate_input_count, estimate_output_count);
+            let total_amount = total_split_amount + estimated_fee;
+            inputs = self
+                .select_outputs(total_amount, 0.into(), UTXOSelectionStrategy::MaturityThenSmallest)
+                .await?;
+            // Check if estimated inputs correct
+            let input_count = inputs.len();
+            if input_count != estimate_input_count {
+                estimate_input_count = input_count;
+                continue;
+            }
+            // Check if estimated outputs correct
+            utxo_total = inputs.iter().fold(MicroTari::from(0), |acc, x| acc + x.value);
+            let output_count = if utxo_total >= total_amount {
+                split_count + 1
+            } else {
+                split_count
+            }; // Might require a change output
+            if output_count != estimate_output_count {
+                estimate_output_count = output_count;
+                continue;
+            }
+            // Check amounts
+            if utxo_total >= total_amount {
+                break;
+            }
+        }
+        let output_count = estimate_output_count;
+
+        trace!(target: LOG_TARGET, "Construct coin split transaction.");
+        let offset = PrivateKey::random(&mut OsRng);
+        let nonce = PrivateKey::random(&mut OsRng);
+        let mut builder = SenderTransactionProtocol::builder(0);
+        builder
+            .with_lock_height(lock_height.unwrap_or(0))
+            .with_fee_per_gram(fee_per_gram)
+            .with_offset(offset.clone())
+            .with_private_nonce(nonce.clone());
+        trace!(target: LOG_TARGET, "Add inputs to coin split transaction.");
+        for uo in inputs.iter() {
+            builder.with_input(
+                uo.as_transaction_input(&self.factories.commitment, uo.clone().features),
+                uo.clone(),
+            );
+        }
+        trace!(target: LOG_TARGET, "Add outputs to coin split transaction.");
+        let mut outputs = Vec::with_capacity(output_count);
+        let change_output = (utxo_total - estimated_fee) - total_split_amount;
+        for i in 0..output_count {
+            let output_amount = if i < output_count - 1 {
+                amount_per_split
+            } else {
+                change_output
+            };
+
+            let mut spend_key = PrivateKey::default();
+            {
+                let mut km = acquire_lock!(self.key_manager);
+                spend_key = km.next_key()?.k;
+            }
+            self.db.increment_key_index().await?;
+            let utxo = UnblindedOutput::new(output_amount, spend_key, None);
+            outputs.push(utxo.clone());
+            builder.with_output(utxo);
+        }
```

Comment:
Is there not a way to do this more succinctly with `tx!` or `txn_schema!`?

Reply:
I attempted to use the `tx!` and `txn_schema!`macro. The `tx!`macro creates new input utxos and doesn't allow the selection of existing utxos. The `txn_schema!` macro can potentially work with some modification, currently it doesn't generate the keys using the correct key manager and it should also return the transaction id. I think in a future PR we can change it to also allow a key manager to be specified and implement a default test key manager when it is used for testing.